ql/src/java/org/apache/hadoop/hive/ql/exec/repl/bootstrap/ReplLoadTask.java

```diff
@@ -0,0 +1,251 @@
+              partitionsTracker);
+          partitionsTracker.debugLog("partitions");
+          break;
+        }
+        case Function: {
+          LoadFunction loadFunction =
+              new LoadFunction(context, (FunctionEvent) next, work.dbNameToLoadIn, dbTracker);
+          TaskTracker functionsTracker = loadFunction.tasks();
+          if (!scope.database) {
+            scope.rootTasks.addAll(functionsTracker.tasks());
+          } else {
+            setUpDependencies(dbTracker, functionsTracker);
+          }
+          loadTaskTracker.update(functionsTracker);
+          functionsTracker.debugLog("functions");
+          break;
+        }
+        }
+      }
+      boolean addAnotherLoadTask = iterator.hasNext() || loadTaskTracker.hasReplicationState();
+      createBuilderTask(scope.rootTasks, addAnotherLoadTask);
+      if (!iterator.hasNext()) {
+        loadTaskTracker.update(updateDatabaseLastReplID(maxTasks, context, scope));
+      }
+      this.childTasks = scope.rootTasks;
+      LOG.info("Root Tasks / Total Tasks : {} / {} ", childTasks.size(), loadTaskTracker.numberOfTasks());
+    } catch (Exception e) {
+      LOG.error("failed replication", e);
+      setException(e);
+      return 1;
+    }
+    LOG.info("completed load task run : {}", work.executedLoadTask());
+    return 0;
+  }
+
+  /**
+   * There was a database update done before and we want to make sure we update the last repl
+   * id on this database as we are now going to switch to processing a new database.
+   */
+  private TaskTracker updateDatabaseLastReplID(int maxTasks, Context context, Scope scope)
+      throws SemanticException {
+  /*
+    we don't want to put any limits on this task as this is essential before we start
+    processing new database events.
+   */
+    TaskTracker taskTracker =
+        new AlterDatabase(context, work.databaseEvent(context.hiveConf), work.dbNameToLoadIn,
+            new TaskTracker(maxTasks)).tasks();
+    scope.rootTasks.addAll(taskTracker.tasks());
+    return taskTracker;
+  }
+
+  private void partitionsPostProcessing(BootstrapEventsIterator iterator,
+      Scope scope, TaskTracker loadTaskTracker, TaskTracker tableTracker,
+      TaskTracker partitionsTracker) throws SemanticException {
+    setUpDependencies(tableTracker, partitionsTracker);
+    if (!scope.database && !scope.table) {
+      scope.rootTasks.addAll(partitionsTracker.tasks());
+      scope.partition = true;
+    }
+    loadTaskTracker.update(tableTracker);
+    loadTaskTracker.update(partitionsTracker);
+    if (partitionsTracker.hasReplicationState()) {
+      iterator.setReplicationState(partitionsTracker.replicationState());
+    }
+  }
+
+  /*
+      This sets up dependencies such that a child task is dependant on the parent to be complete.
+   */
+  private void setUpDependencies(TaskTracker parentTasks, TaskTracker childTasks) {
+    for (Task<? extends Serializable> parentTask : parentTasks.tasks()) {
```

Comment:
This point is true, if we can guarantee that parentTracker have tasks with no children. But, that seems to be an assumption which can be changed in future. So, I feel, it is safer to link dependencies to the tail of parent tasks instead of linking with root.